File: Algorithms/Quick Sort/quicksort.go
package main

import "crypto/rand"

func quicksort(a []int) []int {
	if len(a) < 2 {
		return a
	}

	left, right := 0, len(a)-1

	pivot := rand.Int() % len(a)

	a[pivot], a[right] = a[right], a[pivot]

	for i, _ := range a {
		if a[i] < a[right] {
			a[left], a[i] = a[i], a[left]
			left++
		}
	}

	a[left], a[right] = a[right], a[left]

	quicksort(a[:left])
	quicksort(a[left+1:])

	return a
}

// This function implements the quick sort algorithm in go.
// The quick sort algorithm works by taking an array of values and selecting a pivot value.
// The array is then partitioned such that all values less than the pivot are placed before it
// and all values greater than the pivot are placed after it.
// The array is then recursively sorted with the pivot in its correct position.

File: Algorithms/Quick Sort/quicksort.py
#implement quick sort in python with descriptive comments

def quicksort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quicksort(left) + middle + quicksort(right)

# Quicksort is a divide and conquer algorithm. It works by partitioning an array into two smaller arrays, 
# one containing elements less than the pivot element and one containing elements greater than the pivot element. 
# The algorithm then recursively sorts the two smaller arrays.